package com.kakao.s2graph


import com.stumbleupon.async.{Callback, Deferred}
import play.api.Logger
import play.api.libs.json.JsValue

import scala.concurrent.{Promise, Future}

/**
 * Created by shon on 10/29/15.
 */
package object logger {

  trait Loggable[T] {
    def toLogMessage(msg: T): String
  }

  object Loggable {
    implicit val stringLoggable = new Loggable[String] {
      def toLogMessage(msg: String) = msg
    }

    implicit def numericLoggable[T: Numeric] = new Loggable[T] {
      def toLogMessage(msg: T) = msg.toString
    }

    implicit val jsonLoggable = new Loggable[JsValue] {
      def toLogMessage(msg: JsValue) = msg.toString()
    }
  }

  private val logger = Logger("application")
  private val errorLogger = Logger("error")

  def info[T: Loggable](msg: => T) = logger.info(implicitly[Loggable[T]].toLogMessage(msg))

  def debug[T: Loggable](msg: => T) = logger.debug(implicitly[Loggable[T]].toLogMessage(msg))

  def error[T: Loggable](msg: => T, exception: => Throwable) = errorLogger.error(implicitly[Loggable[T]].toLogMessage(msg), exception)

  def error[T: Loggable](msg: => T) = errorLogger.error(implicitly[Loggable[T]].toLogMessage(msg))
}

package object DeferOp {

  def deferredToFuture[A](d: Deferred[A])(fallback: A): Future[A] = {
    val promise = Promise[A]

    d.addBoth(new Callback[Unit, A] {
      def call(arg: A) = arg match {
        case e: Exception =>
          logger.error(s"deferred failed with return fallback: $e", e)
          promise.success(fallback)
        case _ => promise.success(arg)
      }
    })

    promise.future
  }

  def deferredToFutureWithoutFallback[T](d: Deferred[T]) = {
    val promise = Promise[T]

    d.addBoth(new Callback[Unit, T] {
      def call(arg: T) = arg match {
        case e: Exception =>
          logger.error(s"deferred return Exception: $e", e)
          promise.failure(e)
        case _ => promise.success(arg)
      }
    })

    promise.future
  }

  def deferredCallbackWithFallback[T, R](d: Deferred[T])(f: T => R, fallback: => R) = {
    d.addCallback(new Callback[R, T] {
      def call(args: T): R = {
        f(args)
      }
    }).addErrback(new Callback[R, Exception] {
      def call(e: Exception): R = {
        logger.error(s"Exception on deferred: $e", e)
        fallback
      }
    })
  }

  def errorBack(block: => Exception => Unit) = new Callback[Unit, Exception] {
    def call(ex: Exception): Unit = block(ex)
  }
}
